rust/datafusion/src/optimizer/utils.rs

```diff
@@ -131,116 +131,90 @@ pub fn exprlist_to_fields(expr: &[Expr], input_schema: &Schema) -> Result<Vec<Fi
-
-        (UInt32, UInt8) => Some(UInt32),
-        (UInt32, UInt16) => Some(UInt32),
-        (UInt32, UInt32) => Some(UInt32),
-        (UInt32, UInt64) => Some(UInt64),
-        (UInt32, Float32) => Some(Float32),
-        (UInt32, Float64) => Some(Float64),
-
-        (UInt64, UInt8) => Some(UInt64),
-        (UInt64, UInt16) => Some(UInt64),
-        (UInt64, UInt32) => Some(UInt64),
-        (UInt64, UInt64) => Some(UInt64),
-        (UInt64, Float32) => Some(Float32),
-        (UInt64, Float64) => Some(Float64),
-
-        (Int8, Int8) => Some(Int8),
-        (Int8, Int16) => Some(Int16),
-        (Int8, Int32) => Some(Int32),
-        (Int8, Int64) => Some(Int64),
-        (Int8, Float32) => Some(Float32),
-        (Int8, Float64) => Some(Float64),
-
-        (Int16, Int8) => Some(Int16),
-        (Int16, Int16) => Some(Int16),
-        (Int16, Int32) => Some(Int32),
-        (Int16, Int64) => Some(Int64),
-        (Int16, Float32) => Some(Float32),
-        (Int16, Float64) => Some(Float64),
-
-        (Int32, Int8) => Some(Int32),
-        (Int32, Int16) => Some(Int32),
-        (Int32, Int32) => Some(Int32),
-        (Int32, Int64) => Some(Int64),
-        (Int32, Float32) => Some(Float32),
-        (Int32, Float64) => Some(Float64),
-
-        (Int64, Int8) => Some(Int64),
-        (Int64, Int16) => Some(Int64),
-        (Int64, Int32) => Some(Int64),
-        (Int64, Int64) => Some(Int64),
-        (Int64, Float32) => Some(Float32),
-        (Int64, Float64) => Some(Float64),
-
-        (Float32, Float32) => Some(Float32),
-        (Float32, Float64) => Some(Float64),
-        (Float64, Float32) => Some(Float64),
-        (Float64, Float64) => Some(Float64),
-
-        (Utf8, _) => Some(Utf8),
-        (_, Utf8) => Some(Utf8),
-
-        (Boolean, Boolean) => Some(Boolean),
+    if l == r {
+        return Ok(l.clone());
+    }
 
+    let super_type = match l {
+        UInt8 => match r {
+            UInt16 | UInt32 | UInt64 => Some(r.clone()),
+            Int16 | Int32 | Int64 => Some(r.clone()),
+            Float32 | Float64 => Some(r.clone()),
+            _ => None,
+        },
+        UInt16 => match r {
+            UInt8 => Some(l.clone()),
+            UInt32 | UInt64 => Some(r.clone()),
+            Int8 => Some(l.clone()),
+            Int32 | Int64 => Some(r.clone()),
+            Float32 | Float64 => Some(r.clone()),
+            _ => None,
+        },
+        UInt32 => match r {
+            UInt8 | UInt16 => Some(l.clone()),
+            UInt64 => Some(r.clone()),
+            Int8 | Int16 => Some(l.clone()),
+            Int64 => Some(r.clone()),
+            Float32 | Float64 => Some(r.clone()),
+            _ => None,
+        },
+        UInt64 => match r {
+            UInt8 | UInt16 | UInt32 => Some(l.clone()),
+            Int8 | Int16 | Int32 => Some(l.clone()),
+            Float32 | Float64 => Some(r.clone()),
+            _ => None,
+        },
+        Int8 => match r {
```

Comment:
should we support signed and unsigned with the same width as well? for example (int8, uint8) -> int16.

Reply:
Woudln't (int8, uint8) -> int8 instead of int16? I think signed & unsigned of same length should return signed, so I agree with your question

Reply:
with regards to int8 vs int16, I was thinking 255 from uint8 doesn't fit into range of int8, so increase of width to 16 is required.